float32[] state
int32 done
